test/feature/dynamic_types/DynamicTypesTests.cpp

```diff
@@ -0,0 +1,7670 @@
+    ASSERT_EQ(8u, data->get_item_count());
+    ASSERT_EQ(RETCODE_OK, data->get_int32_values(test2, 0));
+    ASSERT_EQ(test2, Int32Seq({0, 0, 0, 0, 0, 0, 0, 0}));
+    ASSERT_EQ(RETCODE_OK, data->set_int32_values(0, {1, 2, 3, 4, 5, 6, 7, 8}));
+    ASSERT_EQ(RETCODE_OK, data->clear_nonkey_values());
+    ASSERT_EQ(8u, data->get_item_count());
+    ASSERT_EQ(RETCODE_OK, data->get_int32_values(test2, 0));
+    ASSERT_EQ(test2, Int32Seq({0, 0, 0, 0, 0, 0, 0, 0}));
+    ASSERT_EQ(RETCODE_OK, data->set_int32_values(0, {1, 2, 3, 4, 5, 6, 7, 8}));
+    ASSERT_EQ(RETCODE_OK, data->clear_value(1));
+    ASSERT_EQ(RETCODE_BAD_PARAMETER, data->clear_value(100));
+    ASSERT_EQ(8u, data->get_item_count());
+    ASSERT_EQ(RETCODE_OK, data->get_int32_values(test2, 0));
+    ASSERT_EQ(test2, Int32Seq({1, 0, 3, 4, 5, 6, 7, 8}));
+
+    DynamicDataFactory::get_instance()->delete_data(data);
+}
+
+TEST_F(DynamicTypesTests, DynamicType_array_of_arrays)
+{
+    DynamicTypeBuilderFactory::_ref_type factory {DynamicTypeBuilderFactory::get_instance()};
+
+    BoundSeq array_dimensions {{ 2, 2 }};
+    BoundSeq inner_array_dimensions {{ 2 }};
+
+    DynamicTypeBuilder::_ref_type builder {factory->create_array_type(
+                                               factory->get_primitive_type(
+                                                   eprosima::fastdds::dds::TK_INT32), inner_array_dimensions)};
+
+    ASSERT_TRUE(builder);
+    builder = factory->create_array_type(builder->build(), array_dimensions);
+    ASSERT_TRUE(builder);
+    MemberDescriptor::_ref_type member_descriptor {traits<MemberDescriptor>::make_shared()};
+    member_descriptor->type(factory->get_primitive_type(eprosima::fastdds::dds::TK_INT32));
+    member_descriptor->name("Wrong");
+    ASSERT_EQ(builder->add_member(member_descriptor), RETCODE_PRECONDITION_NOT_MET);
+
+    DynamicType::_ref_type created_type {builder->build()};
+    ASSERT_TRUE(created_type);
+
+    DynamicData::_ref_type data {DynamicDataFactory::get_instance()->create_data(created_type)};
+
+    // Test get_member_by_name and get_member_by_index.
+    ASSERT_EQ(MEMBER_ID_INVALID, data->get_member_id_by_name(""));
+    ASSERT_EQ(0, data->get_member_id_at_index(0));
+    ASSERT_EQ(1, data->get_member_id_at_index(1));
+    ASSERT_EQ(2, data->get_member_id_at_index(2));
+    ASSERT_EQ(3, data->get_member_id_at_index(3));
+    ASSERT_EQ(MEMBER_ID_INVALID, data->get_member_id_at_index(4));
+
+    // Test getters, setters and loan_value.
+    ASSERT_NE(data->set_int32_value(MEMBER_ID_INVALID, 10), RETCODE_OK);
+    ASSERT_NE(data->set_string_value(MEMBER_ID_INVALID, ""), RETCODE_OK);
+
+    ASSERT_NE(data->set_uint32_values(0, {1, 2}), RETCODE_OK);
+    ASSERT_EQ(data->set_int32_values(0, {1, 2}), RETCODE_OK);
+    Int32Seq good_seq;
+    ASSERT_EQ(data->get_int32_values(good_seq, 0), RETCODE_OK);
+    ASSERT_EQ(good_seq, Int32Seq({1, 2}));
+
+    auto seq_data = data->loan_value(0);
+    ASSERT_TRUE(seq_data);
```

Comment:
In this case the inner array is treated as a DynamicData internally. Then it can be returned. In the case of single array of primitives, the values are not treated as DynamicData internally, so they cannot be returned in the loan function.